File: backend/app/domain/services/salary_calculator.py
"""Salary Calculator - 급여 계산 오케스트레이터

월급제/시급제를 지원하며 결근 공제 로직을 포함합니다.
"""
from dataclasses import dataclass, field
from decimal import Decimal
from typing import List, Optional

from ..entities import Employee, WorkShift, Allowance
from ..entities.employee import CompanySize
from ..value_objects import Money, WorkingHours
from .insurance_calculator import InsuranceCalculator, InsuranceResult
from .tax_calculator import TaxCalculator, TaxResult
from .overtime_calculator import OvertimeCalculator, OvertimeResult
from .weekly_holiday_pay_calculator import WeeklyHolidayPayCalculator, WeeklyHolidayPayResult
from .absence_calculator import AbsenceCalculator, AbsenceResult


@dataclass
class SalaryCalculationResult:
    """급여 계산 결과"""
    employee: Employee
    base_salary: Money
    allowances: List[Allowance]
    regular_wage: Money
    hourly_wage: Money
    overtime_result: OvertimeResult
    weekly_holiday_result: WeeklyHolidayPayResult
    total_gross: Money
    insurance_result: InsuranceResult
    tax_result: TaxResult
    total_deductions: Money
    net_pay: Money
    wage_type: str = "MONTHLY"
    calculation_month: str = ""
    absence_result: Optional[AbsenceResult] = None


class SalaryCalculator:
    """급여 계산기

    계산 순서:
    1. 기본급 결정 (월급제: 고정급-결근공제 / 시급제: 시급×시간)
    2. 통상시급 계산
    3. 연장/야간/휴일 수당 계산
    4. 주휴수당 계산 (개근 주만)
    5. 총 지급액 → 보험/세금 공제 → 실수령액
    """

    WEEKS_PER_MONTH = Decimal('4.345')  # 365 ÷ 7 ÷ 12

    def __init__(self):
        self.insurance_calculator = InsuranceCalculator()
        self.tax_calculator = TaxCalculator()
        self.overtime_calculator = OvertimeCalculator()
        self.weekly_holiday_calculator = WeeklyHolidayPayCalculator()
        self.absence_calculator = AbsenceCalculator()

    @staticmethod
    def calculate_monthly_regular_hours(
        weekly_hours: int, hours_mode: str = "174"
    ) -> Decimal:
        """월 소정근로시간 동적 계산

        Args:
            weekly_hours: 주 소정근로시간 (scheduled_work_days × daily_work_hours)
            hours_mode: "174" (주휴분리) 또는 "209" (주휴포함)

        Examples:
            40h, 174방식: min(40,40) × 4.345 = 173.8 ≈ 174
            35h, 174방식: min(35,40) × 4.345 = 152.075 ≈ 152
            40h, 209방식: (40 + 40/40×8) × 4.345 = 208.56 ≈ 209
        """
        capped = min(weekly_hours, 40)
        weeks = Decimal('4.345')
        if hours_mode == "209":
            # 주휴시간 포함: (소정근로 + 주휴시간) × 4.345
            weekly_holiday_hours = Decimal(capped) / Decimal('40') * Decimal('8')
            return ((Decimal(capped) + weekly_holiday_hours) * weeks).quantize(
                Decimal('1'), rounding='ROUND_HALF_UP'
            )
        else:
            # 174방식: 소정근로시간만
            return (Decimal(capped) * weeks).quantize(
                Decimal('1'), rounding='ROUND_HALF_UP'
            )

    def calculate(
        self,
        employee: Employee,
        base_salary: Money,
        allowances: List[Allowance],
        work_shifts: List[WorkShift],
        wage_type: str = "MONTHLY",
        hourly_wage_input: int = 0,
        calculation_month: str = "",
        absence_policy: str = "STRICT",
        weekly_hours: int = 40,
        hours_mode: str = "174",
    ) -> SalaryCalculationResult:
        """급여 계산

        Args:
            employee: 근로자 정보
            base_salary: 기본 월급 (월급제)
            allowances: 수당 리스트
            work_shifts: 근무 시프트 리스트 (1개월치)
            wage_type: "MONTHLY" 또는 "HOURLY"
            hourly_wage_input: 시급 (시급제)
            calculation_month: 계산 대상 월 (YYYY-MM)
            absence_policy: 결근 공제 정책
            weekly_hours: 주 소정근로시간 (scheduled_work_days × daily_work_hours)
            hours_mode: "174" (주휴분리) 또는 "209" (주휴포함)
        """
        # 0. 계산월 추론 (미지정 시 시프트 날짜에서)
        if not calculation_month and work_shifts:
            first_date = min(s.date for s in work_shifts)
            calculation_month = first_date.strftime("%Y-%m")

        absence_result = None
        is_over_5 = employee.company_size == CompanySize.OVER_5

        if wage_type == "HOURLY":
            # 시급제: 기본급 = 시급 × 실제 근무시간
            hw = Money(hourly_wage_input)
            total_minutes = sum(
                s.calculate_working_hours().to_minutes()
                for s in work_shifts if not s.is_holiday_work
            )
            total_hours = Decimal(total_minutes) / Decimal('60')
            effective_base = (hw * total_hours).round_to_won()
            hourly_wage = hw
            regular_wage = effective_base
        else:
            # 월급제: 결근 공제 계산
            if calculation_month and work_shifts:
                absence_result = self.absence_calculator.calculate(
                    work_shifts=work_shifts,
                    scheduled_work_days=employee.scheduled_work_days,
                    calculation_month=calculation_month,
                    base_salary=base_salary,
                    absence_policy=absence_policy,
                    is_over_5=is_over_5,
                )
                effective_base = base_salary - absence_result.total_deduction
            else:
                effective_base = base_salary

            # 통상임금 = 기본급(공제 전) + 통상임금 포함 수당
            regular_wage = self._calculate_regular_wage(base_salary, allowances)
            monthly_hours = self.calculate_monthly_regular_hours(weekly_hours, hours_mode)
            hourly_wage = self._calculate_hourly_wage(regular_wage, monthly_hours)

            # 결근 공제 후 주휴수당 계산용 시급 전달
            if absence_result:
                absence_result = AbsenceResult(
                    scheduled_days=absence_result.scheduled_days,
                    actual_work_days=absence_result.actual_work_days,
                    absent_days=absence_result.absent_days,
                    daily_wage=absence_result.daily_wage,
                    wage_deduction=absence_result.wage_deduction,
                    holiday_pay_loss=absence_result.holiday_pay_loss,
                    total_deduction=absence_result.total_deduction,
                    absent_weeks=absence_result.absent_weeks,
                )

        # 3. 연장/야간/휴일 수당
        overtime_result = self.overtime_calculator.calculate(
            work_shifts=work_shifts,
            hourly_wage=hourly_wage,
            company_size=employee.company_size,
            scheduled_work_days=employee.scheduled_work_days,
        )

        # 4. 주휴수당 (개근 주만 지급)
        weekly_holiday_result = self.weekly_holiday_calculator.calculate(
            work_shifts=work_shifts,
            hourly_wage=hourly_wage,
            scheduled_work_days=employee.scheduled_work_days,
        )

        # 5. 총 지급액
        total_gross = self._calculate_total_gross(
            base_salary=effective_base,
            allowances=allowances,
            overtime_pay=overtime_result.total(),
            weekly_holiday_pay=weekly_holiday_result.weekly_holiday_pay,
        )

        # 6. 보험/세금
        taxable_gross = self._calculate_taxable_gross(total_gross, allowances)
        insurance_result = self.insurance_calculator.calculate(taxable_gross)
        tax_result = self.tax_calculator.calculate(
            taxable_gross, employee.dependents_count, employee.children_under_20
        )

        total_deductions = insurance_result.total() + tax_result.total()
        net_pay = total_gross - total_deductions

        return SalaryCalculationResult(
            employee=employee,
            base_salary=effective_base,
            allowances=allowances,
            regular_wage=regular_wage,
            hourly_wage=hourly_wage,
            overtime_result=overtime_result,
            weekly_holiday_result=weekly_holiday_result,
            total_gross=total_gross,
            insurance_result=insurance_result,
            tax_result=tax_result,
            total_deductions=total_deductions,
            net_pay=net_pay,
            wage_type=wage_type,
            calculation_month=calculation_month,
            absence_result=absence_result,
        )

    def _calculate_regular_wage(
        self, base_salary: Money, allowances: List[Allowance]
    ) -> Money:
        """통상임금 = 기본급 + 통상임금 포함 고정수당"""
        regular_wage = base_salary
        for allowance in allowances:
            if allowance.is_regular_wage():
                regular_wage += allowance.amount
        return regular_wage

    def _calculate_hourly_wage(self, regular_wage: Money, monthly_hours: Decimal) -> Money:
        """통상시급 = 통상임금 ÷ 월 소정근로시간"""
        return (regular_wage / monthly_hours).round_to_won()

    def _calculate_total_gross(
        self, base_salary: Money, allowances: List[Allowance],
        overtime_pay: Money, weekly_holiday_pay: Money
    ) -> Money:
        """총 지급액 = 기본급 + 수당 + 가산수당 + 주휴수당"""
        total = base_salary
        for allowance in allowances:
            total += allowance.amount
        total += overtime_pay + weekly_holiday_pay
        return total

    def _calculate_taxable_gross(
        self, total_gross: Money, allowances: List[Allowance]
    ) -> Money:
        """과세 대상 = 총 지급액 - 비과세 수당"""
        non_taxable = Money.zero()
        for allowance in allowances:
            if not allowance.is_taxable:
                non_taxable += allowance.amount
        return total_gross - non_taxable

    @classmethod
    def get_monthly_regular_hours(cls, weekly_hours: int = 40, hours_mode: str = "174") -> Decimal:
        """월 소정근로시간 조회 (동적 계산)"""
        return cls.calculate_monthly_regular_hours(weekly_hours, hours_mode)

File: backend-spring/domain/src/main/kotlin/com/paytools/domain/service/SalaryCalculator.kt
package com.paytools.domain.service

import com.paytools.domain.model.Allowance
import com.paytools.domain.model.CompanySize
import com.paytools.domain.model.Employee
import com.paytools.domain.model.InclusiveWageOptions
import com.paytools.domain.model.InsuranceOptions
import com.paytools.domain.model.WageType
import com.paytools.domain.model.WorkShift
import com.paytools.domain.vo.Money
import com.paytools.domain.vo.WorkingHours
import java.math.BigDecimal
import java.math.RoundingMode

/**
 * 급여 계산 결과
 */
data class SalaryCalculationResult(
    val employee: Employee,
    val baseSalary: Money,
    val allowances: List<Allowance>,
    val regularWage: Money,
    val hourlyWage: Money,
    val overtimeResult: OvertimeResult,
    val weeklyHolidayResult: WeeklyHolidayPayResult,
    val totalGross: Money,
    val insuranceResult: InsuranceResult,
    val taxResult: TaxResult,
    val totalDeductions: Money,
    val netPay: Money,
    val wageType: String = "MONTHLY",
    val calculationMonth: String = "",
    val absenceResult: AbsenceResult? = null,
    val inclusiveWageOptions: InclusiveWageOptions = InclusiveWageOptions.DISABLED,
    val inclusiveOvertimePay: Money = Money.ZERO,
    val appliedWageMode: String? = null,
    val contractVsActualDiff: Money? = null,
    val contractGuaranteeAllowance: Money = Money.ZERO
)

/**
 * 급여 계산 오케스트레이터 (3분류 지원)
 *
 * 급여 유형:
 * - MONTHLY_FIXED: 월급제 고정 (매월 동일, 결근만 공제)
 * - HOURLY_MONTHLY: 시급제 월정산 (실제시간 × 시급)
 * - HOURLY_BASED_MONTHLY: 시급기반 월급제 (MAX(계약월급, 실제계산))
 *
 * 계산 순서:
 * 1. 기본급 결정 (유형별 분기)
 * 2. 통상시급 계산
 * 3. 연장/야간/휴일 수당 계산
 * 4. 주휴수당 계산 (개근 주만)
 * 5. 총 지급액 → 보험/세금 공제 → 실수령액
 */
class SalaryCalculator {

    private val insuranceCalculator = InsuranceCalculator()
    private val taxCalculator = TaxCalculator()
    private val overtimeCalculator = OvertimeCalculator()
    private val weeklyHolidayCalculator = WeeklyHolidayPayCalculator()
    private val absenceCalculator = AbsenceCalculator()

    companion object {
        val WEEKS_PER_MONTH = BigDecimal("4.345")

        /**
         * 월 소정근로시간 동적 계산
         * @param weeklyHours 주 소정근로시간
         * @param hoursMode "174" (주휴분리) 또는 "209" (주휴포함)
         */
        fun calculateMonthlyRegularHours(weeklyHours: Int, hoursMode: String = "174"): BigDecimal {
            val capped = minOf(weeklyHours, 40)
            return if (hoursMode == "209") {
                val weeklyHolidayHours = BigDecimal(capped)
                    .divide(BigDecimal("40"), 10, RoundingMode.HALF_UP)
                    .multiply(BigDecimal("8"))
                (BigDecimal(capped) + weeklyHolidayHours)
                    .multiply(WEEKS_PER_MONTH)
                    .setScale(0, RoundingMode.HALF_UP)
            } else {
                BigDecimal(capped)
                    .multiply(WEEKS_PER_MONTH)
                    .setScale(0, RoundingMode.HALF_UP)
            }
        }
    }

    fun calculate(
        employee: Employee,
        baseSalary: Money,
        allowances: List<Allowance>,
        workShifts: List<WorkShift>,
        wageType: String = "MONTHLY",
        hourlyWageInput: Int = 0,
        calculationMonth: String = "",
        absencePolicy: String = "STRICT",
        weeklyHours: Int = 40,
        hoursMode: String = "174",
        insuranceOptions: InsuranceOptions = InsuranceOptions.ALL_APPLY,
        inclusiveWageOptions: InclusiveWageOptions = InclusiveWageOptions.DISABLED,
        contractMonthlySalary: Long? = null
    ): SalaryCalculationResult {
        // 0. 계산월 추론
        val effectiveMonth = if (calculationMonth.isEmpty() && workShifts.isNotEmpty()) {
            val firstDate = workShifts.minOf { it.date }
            "${firstDate.year}-${firstDate.monthValue.toString().padStart(2, '0')}"
        } else {
            calculationMonth
        }

        val isOver5 = employee.companySize == CompanySize.OVER_5
        val normalizedType = WageType.normalize(WageType.valueOf(wageType))

        // 1. 유형별 기본급/통상시급 결정
        val baseCalcResult = calculateBasePay(
            normalizedType, employee, baseSalary, allowances, workShifts,
            hourlyWageInput, effectiveMonth, absencePolicy, weeklyHours,
            hoursMode, isOver5
        )

        // 2. 연장/야간/휴일 수당
        val (overtimeResult, inclusiveOvertimePay) = calculateOvertimePay(
            normalizedType, inclusiveWageOptions, workShifts,
            baseCalcResult.hourlyWage, employee, baseCalcResult.overtimeForHourly
        )

        // 3. 주휴수당
        val weeklyHolidayResult = weeklyHolidayCalculator.calculate(
            workShifts = workShifts,
            hourlyWage = baseCalcResult.hourlyWage,
            scheduledWorkDays = employee.scheduledWorkDays
        )

        // 4. HOURLY_BASED_MONTHLY: MAX(계약월급, 실제계산) 판정
        val hourlyBasedResult = resolveHourlyBasedMonthly(
            normalizedType, baseCalcResult, weeklyHolidayResult, contractMonthlySalary
        )

        // 5. 총 지급액 (계약보전수당 포함)
        val totalGross = calculateTotalGross(
            baseSalary = hourlyBasedResult.finalBase,
            allowances = allowances,
            overtimePay = overtimeResult.total() + inclusiveOvertimePay,
            weeklyHolidayPay = weeklyHolidayResult.weeklyHolidayPay
        ) + hourlyBasedResult.guaranteeAllowance

        // 6. 보험/세금
        val taxableGross = calculateTaxableGross(totalGross, allowances)
        val insuranceResult = insuranceCalculator.calculate(taxableGross, insuranceOptions)
        val taxResult = taxCalculator.calculate(
            taxableGross, employee.dependentsCount, employee.childrenUnder20
        )

        val totalDeductions = insuranceResult.total() + taxResult.total()
        val netPay = totalGross - totalDeductions

        return SalaryCalculationResult(
            employee = employee,
            baseSalary = hourlyBasedResult.finalBase,
            allowances = allowances,
            regularWage = baseCalcResult.regularWage,
            hourlyWage = baseCalcResult.hourlyWage,
            overtimeResult = overtimeResult,
            weeklyHolidayResult = weeklyHolidayResult,
            totalGross = totalGross,
            insuranceResult = insuranceResult,
            taxResult = taxResult,
            totalDeductions = totalDeductions,
            netPay = netPay,
            wageType = wageType,
            calculationMonth = effectiveMonth,
            absenceResult = baseCalcResult.absenceResult,
            inclusiveWageOptions = inclusiveWageOptions,
            inclusiveOvertimePay = inclusiveOvertimePay,
            appliedWageMode = hourlyBasedResult.appliedWageMode,
            contractVsActualDiff = hourlyBasedResult.contractDiff,
            contractGuaranteeAllowance = hourlyBasedResult.guaranteeAllowance
        )
    }

    /** 유형별 기본급 계산 중간 결과 */
    private data class BaseCalcResult(
        val effectiveBase: Money,
        val hourlyWage: Money,
        val regularWage: Money,
        val absenceResult: AbsenceResult?,
        val overtimeForHourly: OvertimeResult?
    )

    /**
     * Step 1: 유형별 기본급/통상시급 결정
     */
    private fun calculateBasePay(
        normalizedType: WageType, employee: Employee, baseSalary: Money,
        allowances: List<Allowance>, workShifts: List<WorkShift>,
        hourlyWageInput: Int, effectiveMonth: String, absencePolicy: String,
        weeklyHours: Int, hoursMode: String, isOver5: Boolean
    ): BaseCalcResult = when (normalizedType) {

        WageType.MONTHLY_FIXED -> {
            // 월급제 고정: 결근 공제
            var absResult: AbsenceResult? = null
            val effBase: Money
            if (effectiveMonth.isNotEmpty() && workShifts.isNotEmpty()) {
                absResult = absenceCalculator.calculate(
                    workShifts, employee.scheduledWorkDays, effectiveMonth,
                    baseSalary, absencePolicy, isOver5
                )
                effBase = baseSalary - absResult.totalDeduction
            } else {
                effBase = baseSalary
            }
            val regWage = calculateRegularWage(baseSalary, allowances)
            val monthlyHours = calculateMonthlyRegularHours(weeklyHours, hoursMode)
            val hw = calculateHourlyWage(regWage, monthlyHours)
            BaseCalcResult(effBase, hw, regWage, absResult, null)
        }

        WageType.HOURLY_MONTHLY, WageType.HOURLY_BASED_MONTHLY -> {
            // 시급 기반: 소정근로시간 × 시급 (연장시간 제외)
            val hw = Money.of(hourlyWageInput)
            val overtimeForHourly = overtimeCalculator.calculate(
                workShifts, hw, employee.companySize, employee.scheduledWorkDays
            )

            val shiftsForBasePay = if (isOver5) {
                workShifts.filter { !it.isHolidayWork }
            } else {
                workShifts
            }
            val totalMinutes = shiftsForBasePay.sumOf { it.calculateWorkingHours().toMinutes() }
            val overtimeMinutes = if (isOver5) overtimeForHourly.overtimeHours.toMinutes() else 0
            val regularMinutes = totalMinutes - overtimeMinutes
            val regularHours = BigDecimal(regularMinutes).divide(BigDecimal("60"), 10, RoundingMode.HALF_UP)
            val effBase = (hw * regularHours).roundToWon()
            BaseCalcResult(effBase, hw, effBase, null, overtimeForHourly)
        }

        else -> throw IllegalArgumentException("Unknown wage type: $normalizedType")
    }

    /**
     * Step 2: 연장/야간/휴일 수당 (포괄임금제 분기 포함)
     */
    private fun calculateOvertimePay(
        normalizedType: WageType,
        inclusiveWageOptions: InclusiveWageOptions, workShifts: List<WorkShift>,
        hourlyWage: Money, employee: Employee, overtimeForHourly: OvertimeResult?
    ): Pair<OvertimeResult, Money> {
        val isMonthlyFixed = normalizedType == WageType.MONTHLY_FIXED
        if (inclusiveWageOptions.enabled && isMonthlyFixed) {
            val fixedOvertimePay = inclusiveWageOptions.calculateMonthlyFixedOvertimePay()
            val actualResult = overtimeCalculator.calculate(
                workShifts, hourlyWage, employee.companySize, employee.scheduledWorkDays
            )
            val modifiedResult = OvertimeResult(
                overtimePay = Money.ZERO,
                nightPay = actualResult.nightPay,
                holidayPay = actualResult.holidayPay,
                overtimeHours = WorkingHours.fromMinutes(
                    (inclusiveWageOptions.monthlyExpectedOvertimeHours * BigDecimal("60")).toInt()
                ),
                nightHours = actualResult.nightHours,
                holidayHours = actualResult.holidayHours,
                hourlyWage = hourlyWage
            )
            return modifiedResult to fixedOvertimePay
        }

        val result = overtimeForHourly ?: overtimeCalculator.calculate(
            workShifts, hourlyWage, employee.companySize, employee.scheduledWorkDays
        )
        return result to Money.ZERO
    }

    /** HOURLY_BASED_MONTHLY 판정 결과 */
    private data class HourlyBasedResult(
        val finalBase: Money,
        val appliedWageMode: String?,
        val contractDiff: Money?,
        val guaranteeAllowance: Money
    )

    /**
     * Step 4: HOURLY_BASED_MONTHLY 계약월급 vs 실제계산 비교
     * MAX(계약월급, 실제시간×시급+주휴수당) 적용
     * 차액은 기본급에 합산하지 않고 계약보전수당으로 분리
     */
    private fun resolveHourlyBasedMonthly(
        normalizedType: WageType, baseCalcResult: BaseCalcResult,
        weeklyHolidayResult: WeeklyHolidayPayResult,
        contractMonthlySalary: Long?
    ): HourlyBasedResult {
        if (normalizedType != WageType.HOURLY_BASED_MONTHLY || contractMonthlySalary == null) {
            return HourlyBasedResult(baseCalcResult.effectiveBase, null, null, Money.ZERO)
        }

        val contractSalary = Money.of(contractMonthlySalary)
        val actualBase = baseCalcResult.effectiveBase
        val weeklyHolidayPay = weeklyHolidayResult.weeklyHolidayPay
        val actualTotal = actualBase + weeklyHolidayPay

        return if (actualTotal > contractSalary) {
            val diff = actualTotal - contractSalary
            HourlyBasedResult(actualBase, "ACTUAL_CALCULATION", diff, Money.ZERO)
        } else {
            // 계약월급 보장 → 차액을 계약보전수당으로 분리 (기본급 유지)
            val diff = contractSalary - actualTotal
            HourlyBasedResult(actualBase, "CONTRACT_SALARY", diff, diff)
        }
    }

    private fun calculateRegularWage(baseSalary: Money, allowances: List<Allowance>): Money {
        return allowances
            .filter { it.isRegularWage() }
            .fold(baseSalary) { acc, allowance -> acc + allowance.amount }
    }

    private fun calculateHourlyWage(regularWage: Money, monthlyHours: BigDecimal): Money =
        (regularWage / monthlyHours).roundToWon()

    private fun calculateTotalGross(
        baseSalary: Money, allowances: List<Allowance>,
        overtimePay: Money, weeklyHolidayPay: Money
    ): Money {
        val allowanceTotal = allowances.fold(Money.ZERO) { acc, a -> acc + a.amount }
        return baseSalary + allowanceTotal + overtimePay + weeklyHolidayPay
    }

    private fun calculateTaxableGross(totalGross: Money, allowances: List<Allowance>): Money {
        val nonTaxable = allowances
            .filter { !it.isTaxable }
            .fold(Money.ZERO) { acc, a -> acc + a.amount }
        return totalGross - nonTaxable
    }
}
